cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/LocationManager.java

```diff
@@ -0,0 +1,149 @@
+    }
+
+    Location logFile = getLogLocation(locationIdentifier).append(fileName);
+    Location logDir = Locations.getParent(logFile);
+    // check if parent directories exist
+    Locations.mkdirsIfNotExists(logDir);
+
+    if (logFile.exists()) {
+      // The file name for a given application exists if the appender was stopped and then started again but file was
+      // not rolled over. In this case, since the roll over size is typically small, we can rename the old file and
+      // copy its contents to new file and delete old file
+      long now = System.currentTimeMillis();
+      // rename existing file to temp file
+      if (logDir == null) {
+        // this should never happen
+        throw new IOException(String.format("Parent Directory for %s is null", logFile.toURI().toString()));
+      }
+      Location tempLocation = logFile.renameTo(logDir.append(Long.toString(now)));
+      logFile.createNew(filePermissions);
+
+      if (tempLocation == null) {
+        throw new IOException(String.format("Can not rename file %s", logFile.toURI().toString()));
+      }
+
+      OutputStream outputStream = logFile.getOutputStream();
+      try (BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(tempLocation.getInputStream()))) {
+        String line;
+        while ((line = bufferedReader.readLine()) != null) {
+          outputStream.write(line.getBytes());
+        }
+      }
+      // delete temporary file
+      tempLocation.delete();
+      activeLocations.put(locationIdentifier, new LocationOutputStream(logFile, outputStream));
+    } else {
+      // create file with correct permissions
+      logFile.createNew(filePermissions);
+      activeLocations.put(locationIdentifier, new LocationOutputStream(logFile, logFile.getOutputStream()));
+    }
+
+    return activeLocations.get(locationIdentifier).getOutputStream();
+  }
+
+  /**
+   * Closes all open output streams and clears cache
+   */
+  public void close() {
+    Collection<LocationOutputStream> locations = new ArrayList<>(activeLocations.values());
+    activeLocations.clear();
+
+    for (LocationOutputStream locationOutputStream : locations) {
+      // we do not want to throw any exception rather close all the open output streams. so close quietly
+      Closeables.closeQuietly(locationOutputStream);
+    }
+  }
+
+  /**
+   * Flushes all the open output streams
+   */
+  @Override
+  public void flush() throws IOException {
+    Collection<LocationOutputStream> locations = new ArrayList<>(activeLocations.values());
+    for (LocationOutputStream locationOutputStream : locations) {
+      locationOutputStream.flush();
+    }
+  }
+
+  /**
+   * Appends information from location identifier to logBaseDir
+   */
+  Location getLogLocation(LocationIdentifier locationIdentifier) throws IOException {
+    return logBaseDir.append(locationIdentifier.getNamespaceId()).append(locationIdentifier.getApplicationId());
```

Comment:
thats right, the structure of the file name is explained in sample `rolling-appender-logback-test.xml`